### PR TITLE
Created CRUD operations on React Frontend for Bio and Blog functionalities

### PR DESCRIPTION
**Pull request by musanaeem**
Created CRUD functionalities on React Frontend using backend API calls to the CRUD functionalities. 
- Username was stored in local storage to ensure user does not have modification rights to data that does not belong to him. - InputGroup component was made to simplify login/registration form
- Bio and Blog forms were created for create and update
- Create and Update are being done on the same page with conditional components being called
- Data is shared using useContext and useLocation hooks
- Formatting is fixed (indentation, spaces bewteen "=" in JSX and removal of logs)
- Styling is done with specificity to not apply to every page
- Backend modification is done strictly through API calls
- Links pass state which are accessed later by useLocation for specific object identification

Note:
- Refreshing page causes redirect to homepage. Working on a fix to it.

Screenshots:

BlogListPage:
image

BlogPage:
image

BioPage:
image

Update Blog:
image
image

Update Bio:
image
image

Create Bio:
image
image

Delete Bio:
image
image

Delete Blog:
image
image

BlogSite Application is completed.

### CODE REVIEW

File: src/components/BioForm.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            const {data} = state;
+
+        setCredentials({
+            name: data.name,
+            address: data.address,
+            description: data.description
+        })
+    }
+
+    }, []);
```
Comment (srehaider) [inline src/components/BioForm.js]: Can you please explain how this works?
Fix indentation!

File: src/components/Methods.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+  return (
+    <>
+    {
+        props.type === 'Bio' ? getBioMethods(): getBlogMethods()
+    }
+    </>
+  )
+}
```
Comment (srehaider) [inline src/components/Methods.js]:
```suggestion
const BioMethods = (props) => (
    <div className="methods">
        <Link className="method" to='/bio/edit' state={{data: props.data}}> Update </Link>
        <Link className="method" to='/bio/delete'  state={{username: props.data.user_username}}> Delete </Link>
    </div>
);


const BlogMethods = (props) => {
    const {user} = useContext(UserContext);
   
   if (props.data.user_username === user){
        return (
            <div className="methods">
                <Link className="method" to='/blog/edit/:id'  state={{id: props.id, data: props.data}}> Update </Link>
                <Link className="method" to='/blog/delete/:id'  state={{id: props.id, username: props.data.title}}> Delete </Link>
            </div>
        )
    }

    return null;
}

const Methods = (props) => {
    return props.type === 'Bio' ? <BioMethods {...props} /> : <BlogMethods {...props}/>;
}
```

File: src/components/ProtectedAuthentication.js [hunk 3]
```diff
@@ -3,6 +3,7 @@ import { Navigate, Outlet } from 'react-router-dom';
 
 export default class ProtectedAuthentication extends Component {
   render() {
-    return (this.props.isLoggedIn === true) ? <Navigate to='/home'/> : <Outlet/>;
+      return (this.props.isLoggedIn === true) ? <Navigate to='/home'/> : <Outlet/>;
+
```
Comment (srehaider) [inline src/components/ProtectedAuthentication.js]: Why do random new lines are introduced in each PR?

**Review by srehaider — CHANGES_REQUESTED** [review]

File: src/components/BioForm.js (hunk 1 above)
Comment (musanaeem) [reply src/components/BioForm.js]: The update function works as a patch (partial update). The useEffect() runs once after the page has been rendered (Form has been displayed). 

The function checks if the method is patch or post (as the same code runs for the create Bio as well). 

If the method is patch (meaning the data is to be updated), the data that would have been gotten through the state props in Link component ("<Link className="method" to='/blog/edit/:id'  state={{id: props.id, data: props.data}}> Update </Link>"). This data is fetched on another page using the useLocation() hook.

The data is set to the input values of the fields to pre-fill the fields such that the user can choose to update them with the previous values autofilled. The user can leave fields empty as well as the update is a partial update (patch).

The cycle to reach an update is (BioPage --> Record Container --> Method --> Link). The Link component passes the data (as shown above) and calls the UpdateOrCreate Page. That loads the BioForm as an update form and autofills previous data in it.

File: src/components/ProtectedAuthentication.js (hunk 3 above)
Comment (musanaeem) [reply src/components/ProtectedAuthentication.js]: I believe these lines are introduced while I am resolving the conflicts even though I thought I removed as many extra lines as possible. I'll be more careful next time. Apologies.

File: src/components/Methods.js (hunk 2 above)
Comment (musanaeem) [reply src/components/Methods.js]: Changes made as per suggestion

**Review by srehaider — APPROVED** [review]

**Pull request merged**